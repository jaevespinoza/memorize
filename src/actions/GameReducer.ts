// src/features/gameSlice.js
import { createSlice } from "@reduxjs/toolkit";
import { ISelectedImage } from "../interfaces/AppActionsInterface";

export interface IGameState {
  selectedCards: ISelectedImage[];
  foundCards: ISelectedImage[];
  matchPopup: boolean;
  matchedCard: boolean;
  successfulGames: number;
  success: number;
  errors: number;
  name: string;
}

const gameSlice = createSlice({
  name: "gameApplication",
  initialState: {
    selectedCards: [] as ISelectedImage[],
    foundCards: [] as ISelectedImage[],
    success: 0,
    matchPopup: false,
    matchedCard: false,
    errors: 0,
    name: "",
    successfulGames: 0,
  } as IGameState,
  reducers: {
    /**
     * Sets whether the cards match or not
     */
    setMatchedCard: (state, action) => {
      state.matchedCard = action.payload;
    },
    /**
     * Sets whether the match alert should be shown or not
     */
    setMatchedPopup: (state, action) => {
      state.matchPopup = action.payload;
    },
    /**
     * Sets the selected cards into the state and then checks
     * whether they match or not
     */
    setSelectedCards: (state, action) => {
      if (!state.selectedCards.some((item) => item.id === action.payload.id)) {
        state.selectedCards.push(action.payload);
        checkSelected(state);
      }
    },
    /**
     * Sets the found cards into the state so they are still shown
     */
    setFoundCards: (state, action) => {
      state.foundCards.push(action.payload);
    },
    /**
     * Sets the number of successful matches
     */
    setSuccess: (state, action) => {
      state.success = action.payload;
    },
    /**
     * Sets the number of wrong matches
     */
    setErrors: (state, action) => {
      state.errors = action.payload;
    },
    /**
     * Sets the name of the user into the state
     */
    setName: (state, action) => {
      state.name = action.payload;
    },
    /**
     * Resets the selected cards and then adds to the error
     */
    resetCards: (state) => {
      state.selectedCards = [];
    },
    resetGame: (state) => {
      state.foundCards = [];
      state.selectedCards = [];
      state.success = 0;
      state.errors = 0;
      state.successfulGames += 1;
    },
  },
});

/**
 * Checks the cards that were selected.
 * If the cards match, they are added to the pool of found cards and
 * the success popup shows. If not, they are discarded and the
 * error popup shows.
 * @param state The Application state
 * @returns The updated state with the amount of errors and successes updated
 */
const checkSelected = (state: IGameState) => {
  // If the user hasn't selected more than two cards, we return
  if (state.selectedCards.length < 2) {
    return;
  }
  // We compare the names of the cards
  const checkSelectedCards =
    state.selectedCards[0].name === state.selectedCards[1].name;
  state.matchPopup = true;
  // If they match, we set the matchedCard variable to true. If not, it's false.
  if (checkSelectedCards) {
    state.matchedCard = true;
    successOnMatch(state);
  } else {
    state.matchedCard = false;
    state.errors += 1;
  }
};

const successOnMatch = (state: IGameState) => {
  state.foundCards.push(state.selectedCards[0]);
  state.foundCards.push(state.selectedCards[1]);
  state.selectedCards = [];
  state.success += 1;
};

export const {
  setSelectedCards,
  setFoundCards,
  setMatchedCard,
  resetCards,
  resetGame,
  setMatchedPopup,
  setSuccess,
  setErrors,
  setName,
} = gameSlice.actions;
export default gameSlice.reducer;
