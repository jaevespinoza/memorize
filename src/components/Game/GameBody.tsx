import { useSelector } from "react-redux";
import {
  IDuplicatedImage,
  ISelectedImage,
} from "../../interfaces/AppActionsInterface";
import GameCard from "./GameCard";
import "./styles.scss";
import { RootState } from "../../../config/store";
import GameNav from "./GameNav";
import SuccessAlert from "./AlertComponent";
import { useEffect, useRef } from "react";

/**
 * Function that determines whether an id exists on either the selected or found list of cards
 * @param id Id to find
 * @param selected Selected cards at the moment
 * @param found Found cards at the moment
 * @returns Boolean that says whether the id exists
 */
const doesIdExistInArrays = (
  id: number,
  selected: ISelectedImage[],
  found: ISelectedImage[]
) => {
  return (
    selected.some((item) => item.id === id) ||
    found.some((item) => item.id === id)
  );
};

/**
 * Component that renders the various animal cards responsively to
 * play the game.
 * @param data Array of shuffled and repeated cards to play the game
 */
const GameBody = ({ data }: { data: IDuplicatedImage[] }) => {
  const selectedCards = useSelector(
    (state: RootState) => state.game.selectedCards
  );

  const foundCards = useSelector((state: RootState) => state.game.foundCards);

  const cards = useRef([]);
  const gameContainer = useRef(null);

  useEffect(() => {
    // Get all the cards and convert them to an array
    if (cards) {
      cards.current = Array.from(document.querySelectorAll(".card-content"));
      console.log(cards);
    }
  }, [cards]);

  console.log(cards);

  const handleKeyDown = (e) => {
    console.log(e);
    if (e.key === "ArrowRight") {
      e.preventDefault();
      focusNextCard();
    } else if (e.key === "ArrowLeft") {
      e.preventDefault();
      focusPreviousCard();
    }
  };

  const focusNextCard = () => {
    const focusedIndex = cards.current.findIndex(
      (card) => card === document.activeElement
    );
    const nextIndex = (focusedIndex + 1) % cards.current.length;
    cards.current[nextIndex].focus();
  };

  const focusPreviousCard = () => {
    const focusedIndex = cards.current.findIndex(
      (card) => card === document.activeElement
    );
    const previousIndex =
      focusedIndex === 0 ? cards.current.length - 1 : focusedIndex - 1;
    cards.current[previousIndex].focus();
  };

  return (
    <div className="game-container">
      <GameNav />
      <div className="container game-container__body" onKeyDown={handleKeyDown}>
        {data.map((item) => (
          <div
            className="col-lg-1 col-md-2 col-sm-3 col-3 m-1 game-container__card"
            key={item.id}
          >
            <GameCard
              id={item.id}
              revealed={doesIdExistInArrays(item.id, selectedCards, foundCards)}
              url={item.url}
              uuid={item.uuid}
              title={item.title}
              content_type={item.content_type}
            />
          </div>
        ))}
      </div>
      <div className="container notification">
        <SuccessAlert />
      </div>
    </div>
  );
};

export default GameBody;
