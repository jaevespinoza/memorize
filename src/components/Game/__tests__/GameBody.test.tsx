import { configureStore } from "@reduxjs/toolkit";
import { ToolkitStore } from "@reduxjs/toolkit/dist/configureStore";
import { beforeEach, describe, expect, test } from "vitest";
import GameReducer, { IGameState } from "../../../actions/GameReducer";
import GameBody from "../GameBody";
import { render, screen } from "@testing-library/react";
import { Provider } from "react-redux";

describe("Game Body", () => {
  let store: ToolkitStore<{ game: IGameState }>;

  beforeEach(() => {
    store = configureStore({ reducer: { game: GameReducer } });
  });

  test("renders properly", () => {
    const body = render(
      <Provider store={store}>
        <GameBody data={[]} />
      </Provider>
    );

    expect(body).toMatchSnapshot();
  });

  test("renders one card", () => {
    render(
      <Provider store={store}>
        <GameBody
          data={[
            {
              id: 1,
              content_type: "image/jpeg",
              title: "Title",
              url: "",
              uuid: "",
            },
          ]}
        />
      </Provider>
    );

    expect(screen.queryAllByTestId("game-container-card").length).toBe(1);
  });
});
