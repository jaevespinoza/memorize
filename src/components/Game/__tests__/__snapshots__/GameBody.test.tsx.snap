// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Game Body > renders properly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="game-container"
      >
        <nav
          class="navbar navbar-dark bg-primary game-container__header fixed-top"
        >
          <div
            class="container"
          >
            <div
              class="d-flex justify-content-center align-items-center flex-grow-1"
            >
              <h3
                class="game-container__header__title mb-0"
                data-testid="header-title"
              >
                nav.title
              </h3>
            </div>
            <div
              class="game-container__header__counters text-white"
            >
              <span
                class="badge badge-pill badge-success"
                data-testid="success-count"
              >
                nav.success
              </span>
              <span
                class="badge badge-pill badge-danger"
                data-testid="error-count"
              >
                nav.error
              </span>
            </div>
          </div>
        </nav>
        <div
          class="container game-container__body"
          data-testid="game-container-body"
        />
        <div
          class="container notification"
        />
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="game-container"
    >
      <nav
        class="navbar navbar-dark bg-primary game-container__header fixed-top"
      >
        <div
          class="container"
        >
          <div
            class="d-flex justify-content-center align-items-center flex-grow-1"
          >
            <h3
              class="game-container__header__title mb-0"
              data-testid="header-title"
            >
              nav.title
            </h3>
          </div>
          <div
            class="game-container__header__counters text-white"
          >
            <span
              class="badge badge-pill badge-success"
              data-testid="success-count"
            >
              nav.success
            </span>
            <span
              class="badge badge-pill badge-danger"
              data-testid="error-count"
            >
              nav.error
            </span>
          </div>
        </div>
      </nav>
      <div
        class="container game-container__body"
        data-testid="game-container-body"
      />
      <div
        class="container notification"
      />
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
